File: 백준/Gold/12764. 싸지방에 간 준하/싸지방에 간 준하.py
import sys
input = sys.stdin.readline
from heapq import heappop,heappush,nsmallest

n = int(input())
arr = [list(map(int,input().split())) for i in range(n)]
arr.sort()
hq1,hq2=[],list(range(n))
last_time, out_time = arr[0]
heappush(hq1, (out_time, heappop(hq2)))
result = [0]*n
result[0]+=1
max_v = 0
for i in range(1,n):
    a,b =arr[i]
    while hq1 and a>hq1[0][0]:
        _,index =heappop(hq1)
        heappush(hq2,index)
    index = heappop(hq2)
    max_v = max(max_v, index)
    result[index]+=1
    heappush(hq1, (b, index))
    # print(hq1,hq2)
    # print(result)
print(max_v+1)
print(*result[:max_v+1])

# 3
# 1 6
# 2 5
# 3 4

# 3
# 1 1 1

# 3
# 1 11
# 2 10
# 12 13

# ans: 2 1
# result: 1 2

# 5
# 1 5
# 2 3
# 4 5
# 7 8
# 8 9

# # Output
# 2
# 2 3

# 6
# 1 10
# 2 12
# 3 30
# 5 50
# 15 100
# 20 200


# # Answer
# 4
# 2 2 1 1

# 7
# 0 20
# 3 10
# 5 17
# 7 13
# 8 15
# 14 25
# 16 30

# 5
# 1 2 1 2 1

File: 백준/Gold/12764. 싸지방에 간 준하/싸지방에 간 준하.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
	private static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    private static StringTokenizer tokens;
    private static StringBuilder output = new StringBuilder();
    
   
    
    public static void main(String[] args) throws IOException {
    	tokens = new StringTokenizer(input.readLine());
    	int n = Integer.parseInt(tokens.nextToken());
    	int arr[][] = new int[n][2];
    	PriorityQueue<int []> hq1;
    	PriorityQueue<Integer> hq2;
    	int lastTime, outTime;
    	int result[] = new int[n];
    	int index=0;
    	int max_v = 0;
    	for (int i = 0; i < n; i++) {
    		tokens = new StringTokenizer(input.readLine());
    		arr[i][0] = Integer.parseInt(tokens.nextToken());
    		arr[i][1] = Integer.parseInt(tokens.nextToken());
		}
    	Arrays.sort(arr, new Comparator<int[]>() {

			@Override
			public int compare(int[] o1, int[] o2) {
				return o1[0]!=o2[0] ? o1[0]-o2[0] : o1[1]-o2[1];
			}
		});
    	hq1 = new PriorityQueue<>(new Comparator<int []>() {

			@Override
			public int compare(int[] o1, int[] o2) {
				return o1[0]!=o2[0] ? o1[0]-o2[0] : o1[1]-o2[1];
			}
		});
    	hq2 = new PriorityQueue<>();
    	for (int i = 0; i < n; i++) {
			hq2.add(i);
		}
    	
    	
    	
    	lastTime = arr[0][0];
    	outTime = arr[0][1];
    	hq1.add(new int[]{outTime, hq2.poll()});
    	result[0]++;
    	for (int i = 1; i < n; i++) {
    		int a = arr[i][0];
    		int b = arr[i][1];
    		while ((hq1.size()>0) && (a>hq1.peek()[0])) {
				index = hq1.poll()[1];
				hq2.add(index);
			}
			index = hq2.poll();
			
			max_v = Math.max(max_v, index);
			result[index]+=1;
			hq1.add(new int[]{b, index});
		}
    	
    	
    	output.append(max_v+1).append('\n');
    	for (int i = 0; i <= max_v; i++) {
			output.append(result[i]).append(" ");
		}
    	System.out.println(output);
    	
	}
}